### DISCUSSION TITLE
Different ways to read, resolve or update an entity in an array.

### DISCUSSION THREAD

**Discussion by manojsethi** [urql-graphql/urql#1918] [Q&A] [answered]
Hi Team, Just wanted to learn more and understand various ways to update an entity inside an array.
From the documentation, I can see that read and write fragment is the easiest way but still wanted to know if we can somehow do updates via updateQuery.

Taking the example of Ben's course let's say we have a PaginatedPost entity and it has a further Posts array. Now if I want to resolve a particular entity via cache.resolve how can I achieve that.

Once the posts are fetched from the graphql server if my understanding is correct in graphcache it will be normalized and each post will be saved in cache with it's id being the part of the key

So I tried to do a readQuery after pagination data is fetched.

```
cache.readQuery({
                  query: gql`
                    query ($id: ID!) {
                      post(id: $id) {
                        id
                        points
                        voteStatus
                      }
                    }
                  `,
                  variables: {
                    id: args.postId,
                  },
                })
```
                
but it returns null.

I understand the process of getting a particular field via cache.resolve, with chaining I can resolve a step further as well but when there is a case of array entities how can I resolve a particular entity in a list of entities.

 
In same way how can I update the cache for a particular entity. 

I have achieved success via Ben's way of reading and write fragment but wanted to know if there are other ways even if they are not better than read and write fragments.

**Comment by kitten** [comment]
This is quite a straightforward explanation, I hope ✌️

Say, you run `Query.post(id: $id)` and you've got a different field like `Query.posts` that lists all posts. Let's say we never actually query the first field in our UI. We now have a situation where with your update Graphcache doesn't know two pieces of information:

- It doesn't know that both fields relate to "post(s)" (althoug with a schema it could, however...)
- It also does not know that the `id` argument in `Query.post(id: $id)` can be used to resolve a Post.

It simply can't make enough assumptions about your schema to do so. This is why fragments work. With a fragment on a Post you skip this field that Graphcache doesn't know about.

However, we can teach Graphcache how to resolve this field: https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#resolving-entities

If you add a resolver for the `post` field, Graphcache will know how to get to a Post directly from `Query.post` by using the `id` argument:

```js
cacheExchange({
  resolvers: {
    Query: {
      post: (_, args) => ({ __typename: 'Post', id: args.id }),
    },
  },
});
```

This will also have the benefit that if you then did this query in your UI, if the cache already has enough data, even if it has never run a query with this field before, it is possibly able to retrieve the Post directly. ☀️

**Comment by manojsethi** [comment]
Thanks for the explanation, I will try now and then mark the answer. Hoping for success.

**Comment by manojsethi** [comment]
Hi @kitten if I write a resolver like above and in mutation update I want to read the value it still brings null for the above readQuery statement. 

```
 cache.readQuery({
                  query: gql`
                    query ($id: ID!) {
                      post(id: $id) {
                        id
                        points
                        voteStatus
                      }
                    }
                  `,
                  variables: {
                    id: args.postId,
                  },
                })
```

prints null. I can confirm that the post with that particular postid is rendered in ui. Kindly guide.

**Reply by kitten** [reply]
Oh, sorry, I just remembered there was a bit missing in my explanation; I just realised you were talking about cache updates. During cache updates, we actually skip resolvers. That's because if you have a resolver that _changes_ values (let's say it converts date strings to `new Date()` objects) we'd run the risk of making it easier to accidentally write this to the cache, which would then break as the type isn't as expected anymore.

So, for updates `Query.post` won't be resolved, even if you add a resolver if it's unknown. So, I guess the conclusion here is: You have to use fragments and the corresponding read/write fragment APIs if you're updating a specific one, otherwise you can use queries & resolvers in... well, resolvers 😅 Sorry for the added confusion!

It would be helpful to know however what you're trying to update.

> learn more and understand various ways to update an entity inside an array.

Because if this is the question, as you're understanding already "normalisation" means that each entity from your queries is stored by a key, which means that you'll in theory never have to update an entity manually. It's mostly relations (aka "links") that may need to be updated manually, occasionally, if they're not part of a mutation you may have run.

**Comment by manojsethi** [comment]
My basic requirement is to get the post entity in a mutation which I am currently getting via the readFragment way. I don't know if you have ever watched the ben's course as you are already the author of urql and hence expert, I am basically doing upvoting to a post and want to update its points. If I have listened carefully Ben told that I have shown you some other ways of updating the cache earlier and you can use those methods to update the post entity via those methods but this time I will use readFragment. Earlier methods that he had shown was of readQuery and updateQuery. 

So I was just trying to get the things done in the old fashion thinking of getting a better understanding of the cache. But honestly, I was not able to drill down to my post via post id in the mutation and hence thought of learning from the community. I am glad you from your busy schedule are helping us.

**Reply by kitten** [reply]
> But honestly, I was not able to drill down to my post via post id in the mutation

If you post a snippet, I'm happy to give you a hand :v: Most of the time, a couple of `cache.resolve` calls can be used to deeply access data. That's basically an operation that's like accessing deep properties on the cache itself. Once you then arrive at the todo you'll be able to use the resolved key to update it.

**Reply by manojsethi** [reply]
I was able to get the keys of the post via the following resolve

let myPostKeys = cache.resolve(cache.resolve({ __typename: "Query" }, "posts", { limit: 10 }) as string, "posts") as string[]

but getting no clues to get a particular post and update that after the upper resolve.

**Reply by kitten** [reply]
Ok, good so far :+1: that `keys` array now contains keys, but if you're using `readFragment` you'll need to have your keyable fields, say:

```js
const keys = cache.resolve(cache.resolve("Query", "posts", { limit: 10 }) as string, "posts") as string[];

for (const key of keys) {
  const id = cache.resolve(key, "id"); // Get the id field of the post
  if (id) {
    const post = cache.readFragment(
      gql`fragment _ on Post { id yourField }`,
      { id: cache.resolve(keys }
    );
  }
}
```

**Reply by manojsethi** [reply]
I am sorry if I was unclear in the starting post. I want to know that in such a case is there any other way apart from readFragment to reach out till my post entity. When posts array is fetched then a post is stored as post:keyid, so can't  cache.readQuery reach out to that node or cache.readQuery works only for the queries that has been executed and are available at root Query.

**Reply by manojsethi** [reply]
@kitten If my question is wrong or if doesn't make sense to you kindly let me know, I will try to explain better.

**Comment by manojsethi** [comment, marked as answer]
@kitten Sorry to bother you again, can you just highlight the difference on when to use cache.readQuery and cache.readFragment from the documentation I am not getting an idea.

Is my following assumption correct?

cache.readQuery works only when the queries has already been executed via UI and it's key is available under Query ie if the following code brings has a node for that query

`cache.inspectFields({ __typename: 'Query' });`

**Reply by kitten** [reply]
Hiya 👋

That's right. Inside the `updates` config, I'd say you'd want to access the "smallest" amount of data as a rule of thumb, so I'd recommend `readFragment`, `writeFragment`, `cache.resolve` and `cache.link` as needed, unless you're looking to just invalidate data of course.

Reading queries will work in so far as to whether all the fields you're querying are known already, so it's only useful when writing an updater for queries that have already run. That's useful for taking care of simple, individual screens and their queries.

Oftentimes there are three different cases that you're taking care of with updaters:

- Invalidating data (deleting data from the cache and hence triggering refetches automatically)
- Updating relations (which we call links between entities), like a field on one object that contains another
- Updating records or individual fields (like a counter)

The second case here is important when you have a mutation that won't resolve as much data as you need. Think of a `addTagToBook(bookId: ID!, tagName: String!): Tag` mutation for instance. With such a field there isn't a way for the normalised cache to know that the `Book` with the `bookId` has changed. This is when we'd want to use `cache.resolve` and `cache.link` for the shortest amount of code.

For the third case, let's say we've got a mutation that does `upvoteBook(bookId: ID!): User` mutation. If this mutation doesn't resolve the `Book`, we'd need to use `writeFragment` to manually update an `upvoteCount: Int` field.

In both of these cases it'd be easier to have a GraphQL mutation that resolves the object that has changed. If we had `addTagToBook(bookId: ID!, tagName: String!): Book` instead, then we'd already get the updated `Book` and hence the relation.
If we have `upvoteBook(bookId: ID!): Book` then we'd already be able to get the updated `Book` with a new `upvoteCount` field.

In other words, updaters need to be added whenever the Mutation fields we use don't give us enough information about what has changed themselves.

The extreme of this are lists. It's often hard to update all lists automatically. For instance, if we have a user's favourite books, which we query on a special field, and this list may be thousands of items long, and it's sorted alphabetically, then we likely have no chance to update this list, because we may not even have the full list. That's when invalidation is more useful.
For shorter lists it's likely that updating a query that we know is affected would be more efficient. This may happen in eCommerce apps that have shopping baskets for instance, because those lists are often short and easily updatable. And again, that's only needed if the hypothetical mutation that puts new items in the cart doesn't resolve the new cart in our API's schema.

So, to summarise, I'd recommend to think of writing the shortest updater that gets the job done, if possible. In top of that, you'll typically write updaters that overcome schema or information limitations.
So the hierarchy is:

- Can I use `cache.link` to update a relation?
- Can I use a fragment to update a field?
- Can I use a fragment to update a list?
- Can I use a query to update a list?
- Can I invalidate the affected data?

If you go through these questions in this order, you're very likely to write the shortest updater as needed, without adding too much extra code.
This of course changes a bit if you control the schema and API. Sometimes it's easier to adapt the schema to give your app (and hence the normalised cache) the data that Mutations affect directly.

**Reply by manojsethi** [reply]
🙏Thank You. I owe a lot to you for the time spent. 🙏

Is this project there on opencollective?
